Vue Support - #83

zulus opened the pull request:
I wasn't sure howto correctly register other language files defined in https://github.com/vuejs/language-tools/blob/master/packages/vscode-vue/package.json

*.java classes are mostly copy-paste from typescript server

zulus commented:
what now?

vrubezhny commented:
> what now?

@zulus Thanks for contributing! 

It would be great  if you add at least some simple JUnit tests with vue.js cases (like to see if vue.js specific diagnostic/code assist/etc.  is provided by the LS),

vrubezhny commented:
/request-license-review

vrubezhny commented:
/request-license-review

vrubezhny commented:
The IP Team Review requests are created regarding the proposal:

- [npm/npmjs/@vue/language-core/1.7.0](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8796)
- [npm/npmjs/-/volar-service-emmet/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8795)
- [npm/npmjs/-/volar-service-html/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8794)
- [npm/npmjs/@vscode/emmet-helper/2.8.9](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8793)
- [npm/npmjs/-/typesafe-path/0.2.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8792)
- [npm/npmjs/-/volar-service-css/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8791)
- [npm/npmjs/-/volar-service-pug/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8790)
- [npm/npmjs/-/volar-service-typescript/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8789)
- [npm/npmjs/-/volar-service-typescript-twoslash-queries/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8788)
- [npm/npmjs/@johnsoncodehk/pug-beautify/0.2.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8787)
- [npm/npmjs/-/volar-service-json/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8786)
- [npm/npmjs/@vue/language-service/1.7.0](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8785)
- [npm/npmjs/@babel/parser/7.22.5](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8784)
- [npm/npmjs/-/volar-service-pug-beautify/0.0.2](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8783)
- [npm/npmjs/@vue/language-server/1.7.0](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/8782)

zulus commented:
> > what now?
> 
> @zulus Thanks for contributing!
> 
> It would be great if you add at least some simple JUnit tests with vue.js cases (like to see if vue.js specific diagnostic/code assist/etc. is provided by the LS),

Sure!

zulus commented:
I prepared basic tests, similar to angular:

1. Test marker inside method
2. Test marker inside template
3. Completion proposal based on component javascript source
4. Completion proposal based on broken template

vrubezhny commented:
> BTW: during playing with tests, I realized that TestAngular and TestESLint might use system node, rather than built-in

Why do you think so? They both use `NodeJSManager.getNpmLocation().getAbsolutePath()` to get an `npm` executable location, then add arguments for options.

In case of your test, you're generating a command that runs Node.js executable, which gets a path to NPM executable as first argument, then arguments for options... 
```
[D:\a\wildwebdeveloper\wildwebdeveloper\org.eclipse.wildwebdeveloper.tests\target\work\.node\node-v18.16.0-win-x64\node.exe, D:\a\wildwebdeveloper\wildwebdeveloper\org.eclipse.wildwebdeveloper.tests\target\work\.node\node-v18.16.0-win-x64\npm.cmd, install, --no-bin-links, --ignore-scripts
```
not sure how this supposed to work...

zulus commented:
> > BTW: during playing with tests, I realized that TestAngular and TestESLint might use system node, rather than built-in
> 
> Why do you think so? They both use `NodeJSManager.getNpmLocation().getAbsolutePath()` to get an `npm` executable location, then add arguments for options.
> 
> In case of your test, you're generating a command that runs Node.js executable, which gets a path to NPM executable as first argument, then arguments for options...
> 
> ```
> [D:\a\wildwebdeveloper\wildwebdeveloper\org.eclipse.wildwebdeveloper.tests\target\work\.node\node-v18.16.0-win-x64\node.exe, D:\a\wildwebdeveloper\wildwebdeveloper\org.eclipse.wildwebdeveloper.tests\target\work\.node\node-v18.16.0-win-x64\npm.cmd, install, --no-bin-links, --ignore-scripts
> ```
> 
> not sure how this supposed to work...

On mac, this code produces me only path to NPM location. NPM for *nix have `/bin/env node`, but tests ignore PATH produced by NodeJSManager

vrubezhny commented:
> , on *nix will return true, `+x` should be removed from npm script files, or just remove "can execute" checking for npm.

I wonder why nobody else did complain about Mac & Win NPM run failures until now...
Unfortunately I cannot verify in locally on Mac, but previously ran tests for Angular and ESLint stuff successfully execute NPM install using the embedded NPM executable.

If it's true then it's worth reporting it as an issue

vrubezhny commented:
(Sorry, closed unintentionally by clicking the wrong button. Everything is restored and open)

vrubezhny commented on org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/VueClientImpl.java:
```diff
@@ -0,0 +1,28 @@
+package org.eclipse.wildwebdeveloper.vue;
```
Please add license header here

vrubezhny commented on org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/VueLanguageServerExtention.java:
```diff
@@ -0,0 +1,17 @@
+package org.eclipse.wildwebdeveloper.vue;
```
Please add License header here

vrubezhny reviewed (changes requested):
Overall looks good. Please see the requested changes to be fixed before we can merge

vrubezhny commented:
@zulus  Could you please rebase and resolve the requested changes?

zulus commented:
> @zulus Could you please rebase and resolve the requested changes?

Done, I updated license headers, added missing one, and rebased to current HEAD (I see license problems now :/)

BTW, in the meantime vue-language server has been upgraded (1.7.0 to 1.8.3 - https://github.com/vuejs/language-tools/blob/master/CHANGELOG.md), so after merge I'll prepare another pull request with new version

vrubezhny commented:
/request-license-review

vrubezhny commented:
> Done, I updated license headers, added missing one, and rebased to current HEAD (I see license problems now :/)

I created the Review requests for the updated modules, hope they will be resolved soon.

> BTW, in the meantime vue-language server has been upgraded (1.7.0 to 1.8.3 - https://github.com/vuejs/language-tools/blob/master/CHANGELOG.md), so after merge I'll prepare another pull request with new version

Actually you can add this change right now - it will require an additional review for the LS (as it's minor version is changed), but we haven't had any problems with the IP Team review for the previous version, so I don't expect any with the latest one.

zulus commented:
> Actually you can add this change right now - it will require an additional review for the LS (as it's minor version is changed), but we haven't had any problems with the IP Team review for the previous version, so I don't expect any with the latest one.

Done

vrubezhny commented:
/request-license-review

vrubezhny commented on org.eclipse.wildwebdeveloper.tests/testProjects/vue-app/package-lock.json:
Please remove `package-lock.json` - it will be generated as needed during the test project installation.

zulus replied:
done

vrubezhny commented:
I've started the IP Team review for Vue.js LS 1.8.3 - so need to wait until it finshes

@zulus please follow https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/pull/1241#discussion_r1245173162

vrubezhny reviewed (approved):
The code looks good to me.

@zulus Thank you for the contribution!
